### PR TITLE
core: make ManagedChannel honor Service config LB

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This is a roll forward of #3889

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        }
+      }
+
+      String serviceConfigChoiceBalancingPolicy =
+          ServiceConfigUtil.getLoadBalancingPolicy(config);
+
+      // Check for an explicitly present lb choice
+      if (serviceConfigChoiceBalancingPolicy != null) {
+        if (serviceConfigChoiceBalancingPolicy.toUpperCase(Locale.ROOT).equals("ROUND_ROBIN")) {
+          ClassNotFoundException caught = null;
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: unused

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+          try {
+            Class<?> lbFactoryClass;
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: unnecessary separation of declaration and assignment.

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+     *
+     * @param servers The list of servers reported
+     * @param config the service config object
+     * @return the new load balancer factory, or null if the existing lb should be used.
+     */
+    @Nullable
+    @VisibleForTesting
+    static LoadBalancer.Factory decideLoadBalancerFactory(
+        List<EquivalentAddressGroup> servers, Map<String, Object> config) {
+
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: nit: remove one empty line

Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: Done

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java (hunk 1 above)
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: Done.

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java (hunk 2 above)
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: Done.

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+          }
+        }
+        throw new IllegalArgumentException(
+            "Unknown service config policy: " + serviceConfigChoiceBalancingPolicy);
+      }
+
+      return PickFirstBalancerFactory.getInstance();
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java]: Done

File: core/src/test/java/io/grpc/internal/AutoConfiguredLoadBalancerFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      super(null);
+    }
+  }
+
+  private static class TestHelper extends ForwardingLoadBalancerHelper {
+    TestHelper() {
+      super(null);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply core/src/test/java/io/grpc/internal/AutoConfiguredLoadBalancerFactoryTest.java]: Done.

**Pull request merged**